tournamentcontrol/competition/tests/test_utils.py

```diff
@@ -264,25 +265,217 @@ def test_edit_stage_view_with_undecided_teams_referencing_missing_pools(self):
+        # Test that accessing choices returns division teams as fallback
         with self.subTest(property="choices", formula="G1P1"):
-            with self.assertRaisesRegex(
-                IndexError, r"Invalid group 1 for stage .* \(stage has 0 pools\)"
-            ):
-                _ = undecided_team_1.choices  # This also accesses stage_group_position
+            choices = undecided_team_1.choices  # This should return division teams
+            self.assertEqual(choices, stage2.division.teams)
+
+
+class TwoStageFormulaProblemTests(TestCase):
+    """
+    Tests for the two-stage scenario as requested by @goodtune.
+
+    This covers:
+    - Two stages (Round Robin, Finals)
+    - Stage 1 (Round Robin) has two pools
+    - Stage 2 (Finals) has matches with home_team_eval and away_team_eval
+      defined in terms of group and position from the earlier stage
+    - Positive test: valid formulae like G1P1 vs G2P2
+    - Negative test: valid formulae that can't be resolved like G1P1 vs G5P2
+    - When invalid formula occurs, title should return the formula instead of raising exception
+    """
+
+    def setUp(self):
+        """Set up the two-stage competition scenario."""
+        # Create superuser for admin access
+        self.user = SuperUserFactory.create()
+
+        # Create a division with two stages
+        self.division = DivisionFactory.create()
+
+        # Stage 1: Round Robin with two pools
+        self.round_robin_stage = StageFactory.create(
+            division=self.division, order=1, title="Round Robin"
+        )
+
+        # Create two pools in the Round Robin stage
+        self.pool1 = StageGroupFactory.create(stage=self.round_robin_stage, order=1)
+        self.pool2 = StageGroupFactory.create(stage=self.round_robin_stage, order=2)
+
+        # Stage 2: Finals stage that follows Round Robin
+        self.finals_stage = StageFactory.create(
+            division=self.division,
+            order=2,
+            title="Finals",
+            follows=self.round_robin_stage,
+        )
+
+    def test_valid_formulas_resolve_correctly(self):
+        """
+        Positive test: Valid formulae like G1P1 vs G2P2 should work correctly.
+        """
+        # Create UndecidedTeam objects with valid formulas
+        undecided_team_g1p1 = UndecidedTeamFactory.create(
+            stage=self.finals_stage,
+            formula="G1P1",  # Winner of Pool 1, Position 1
+            label="Winner Pool 1",
+        )
+
+        undecided_team_g2p2 = UndecidedTeamFactory.create(
+            stage=self.finals_stage,
+            formula="G2P2",  # Runner-up of Pool 2, Position 2
+            label="Runner-up Pool 2",
+        )
+
+        # Test that the UndecidedTeam.title properties work correctly
+        # Valid formulas should not return the raw formula string
+        self.assertNotEqual(undecided_team_g1p1.title, "G1P1")
+        self.assertNotEqual(undecided_team_g2p2.title, "G2P2")
```

Comment:
This is still not asserting on what we expect, just confirming that it isn't the formula.

We get the best result if we prove that the `title` attribute is "1st Pool A" and "2nd Pool B".

Reflect this positive outcome testing style across all scenarios, don't settle for "not failing".